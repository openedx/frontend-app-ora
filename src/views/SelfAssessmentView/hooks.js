import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

const useAssessmentViewHooks = () => {
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  return {
    submission: pageData?.submission,
    oraConfigData,
  };
};

export default useAssessmentViewHooks;
