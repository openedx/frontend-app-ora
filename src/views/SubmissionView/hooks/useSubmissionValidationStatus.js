import { useCallback } from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useSubmissionConfig,
} from 'hooks/app';

export const stateKeys = StrictDict({
  promptStatuses: 'promptStatuses',
  submissionTriggered: 'submissionTriggered',
  fileUploadIsRequired: 'fileUploadIsRequired',
});

const useSubmissionStatuses = (textResponses, uploadedFiles) => {
  const submissionConfig = useSubmissionConfig();

  // holds the status of each text field
  const [promptStatuses, setPromptStatuses] = useKeyedState(
    stateKeys.promptStatuses,
    textResponses.reduce((obj, val, i) => ({ ...obj, [i]: true }), {}),
  );
  const [submissionTriggered, setSubmissionTriggered] = useKeyedState(stateKeys.submissionTriggered, false);
  const [fileUploadIsRequired, setFileUploadIsRequired] = useKeyedState(stateKeys.fileUploadIsRequired, false);

  // this function is to be called when response submit button is clicked
  const validateBeforeConfirmation = useCallback(() => {
    setSubmissionTriggered(true);
    const retrievePromptsSubmissionStatus = textResponses.reduce((obj, val, i) => ({ ...obj, [i]: val }), {});
    let containsInvalidPrompts = false;
    if (submissionConfig.textResponseConfig.required) {
      for (let i = 0; i < textResponses.length; i++) {
        retrievePromptsSubmissionStatus[i] = !!textResponses[i];
      }
      containsInvalidPrompts = Object.values(retrievePromptsSubmissionStatus).some((val) => !val);
    }
    const calcFileUploadIsRequired = submissionConfig.fileResponseConfig.required && !uploadedFiles.length;
    setFileUploadIsRequired(calcFileUploadIsRequired);
    if (containsInvalidPrompts || (submissionConfig.fileResponseConfig.required && !uploadedFiles.length)) {
      setPromptStatuses(retrievePromptsSubmissionStatus);
      return false;
    }

    return true;
  }, [
    textResponses,
    uploadedFiles,
    submissionConfig,
    setPromptStatuses,
    setSubmissionTriggered,
    setFileUploadIsRequired,
  ]);

  return {
    validateBeforeConfirmation,
    setSubmissionTriggered,
    submissionTriggered,
    setPromptStatuses,
    promptStatuses,
    fileUploadIsRequired,
  };
};

export default useSubmissionStatuses;
