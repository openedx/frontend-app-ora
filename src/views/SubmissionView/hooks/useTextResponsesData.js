import { useCallback, useEffect } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import {
  useFinishLater,
  useHasSubmitted,
  useSaveDraftResponse,
  useTextResponses,
} from 'hooks/app';
import { MutationStatus } from 'constants/index';

export const stateKeys = StrictDict({
  textResponses: 'textResponses',
  isDirty: 'isDirty',
  hasSaved: 'hasSaved',
  lastChanged: 'lastChanged',
});

const useTextResponsesData = ({ setHasSavedDraft }) => {
  const textResponses = useTextResponses();
  const hasSubmitted = useHasSubmitted();
  const [hasSaved, setHasSaved] = useKeyedState(stateKeys.hasSaved, false);
  const [lastChanged, setLastChanged] = useKeyedState(stateKeys.lastChanged, null);
  const [isDirty, setIsDirty] = useKeyedState(stateKeys.isDirty, false);
  const [value, setValue] = useKeyedState(stateKeys.textResponses, textResponses);

  const saveResponseMutation = useSaveDraftResponse();
  const finishLaterMutation = useFinishLater();

  const saveResponse = useCallback(() => {
    setIsDirty(false);
    return saveResponseMutation.mutateAsync({ textResponses: value }).then(() => {
      setHasSavedDraft(true);
    });
  }, [
    saveResponseMutation,
    value,
    setIsDirty,
    setHasSavedDraft,
  ]);

  const finishLater = useCallback(() => {
    setIsDirty(false);
    return finishLaterMutation.mutateAsync({ textResponses: value });
  }, [setIsDirty, finishLaterMutation, value]);

  const onChange = useCallback((index) => (textResponse) => {
    setValue(oldResponses => {
      const out = [...oldResponses];
      out[index] = textResponse;
      return out;
    });
    setIsDirty(true);
    setLastChanged(Date.now());
  }, [
    setValue,
    setIsDirty,
    setLastChanged,
  ]);

  useEffect(() => {
    const interval = setInterval(() => {
      if (!lastChanged) {
        return;
      }
      const timeSinceChange = Date.now() - lastChanged;
      if (isDirty && timeSinceChange > 2000) {
        saveResponse();
        if (!hasSaved) {
          setHasSaved(true);
        }
      }
    }, 2000);
    return () => clearInterval(interval);
  }, [
    isDirty,
    hasSubmitted,
    hasSaved,
    setHasSaved,
    lastChanged,
    saveResponse,
  ]);

  return {
    textResponses: value,
    onUpdateTextResponse: onChange,
    isDraftSaved: saveResponseMutation.status === MutationStatus.success && !isDirty,
    saveResponse,
    saveResponseStatus: saveResponseMutation.status,
    finishLater,
    finishLaterStatus: finishLaterMutation.status,
  };
};

export default useTextResponsesData;
