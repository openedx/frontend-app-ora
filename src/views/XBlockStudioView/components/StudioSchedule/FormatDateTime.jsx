import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import messages from '../messages';

const FormatDateTime = ({ date }) => {
  const { formatMessage, formatDate } = useIntl();
  return (
    <span>{date ? formatDate(date, {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
    }) : formatMessage(messages.notSetLabel)}
    </span>
  );
};

FormatDateTime.defaultProps = {
  date: null,
};
FormatDateTime.propTypes = {
  date: PropTypes.string,
};

export default FormatDateTime;
