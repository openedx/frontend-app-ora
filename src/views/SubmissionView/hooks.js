import { useEffect, useReducer } from 'react';

import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

import { submitResponse, saveResponse, uploadFiles } from 'data/services/lms/hooks/actions';
import { MutationStatus } from 'data/services/lms/constants';

const useSubmissionViewHooks = () => {
  const submitResponseMutation = submitResponse();
  const saveResponseMutation = saveResponse();
  const uploadFilesMutation = uploadFiles();
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  const [submission, dispatchSubmission] = useReducer(
    (state, payload) => ({ ...state, isDirty: true, ...payload }),
    { ...pageData?.submission, isDirty: false },
  );

  useEffect(() => {
    // a workaround to update the submission state when the pageData changes
    if (pageData?.submission) {
      dispatchSubmission({ ...pageData.submission, isDirty: false });
    }
  }, [pageData?.submission]);

  const onTextResponseChange = (index) => (textResponse) => {
    dispatchSubmission({
      response: {
        ...submission.response,
        textResponses: [
          ...submission.response.textResponses.slice(0, index),
          textResponse,
          ...submission.response.textResponses.slice(index + 1),
        ],
      },
    });
  };

  const onFileUploaded = (args) => {
    const fileUploads = uploadFilesMutation.mutate(args);
    dispatchSubmission({ response: { ...submission.response, fileUploads } });
  };

  const submitResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    submitResponseMutation.mutate(submission);
  };

  const saveResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    saveResponseMutation.mutate(submission);
  };

  return {
    submitResponseHandler,
    submitResponseStatus: submitResponseMutation.status,
    saveResponseHandler,
    saveResponseStatus: saveResponseMutation.status,
    draftSaved: saveResponseMutation.status === MutationStatus.success && !submission.isDirty,
    pageData,
    oraConfigData,
    submission,
    dispatchSubmission,
    onTextResponseChange,
    onFileUploaded,
  };
};

export default useSubmissionViewHooks;
