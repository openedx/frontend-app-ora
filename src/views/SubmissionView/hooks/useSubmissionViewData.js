import React from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useGlobalState,
  useRubricConfig,
  useSubmitResponse,
  useSetHasSubmitted,
  useHasSubmitted,
  useRefreshPageData,
  useSetResponse,
  useResponse,
} from 'hooks/app';
import {
  useRefreshUpstream,
} from 'hooks/modal';
import { stepStates, stepNames } from 'constants/index';

import useTextResponsesData from './useTextResponsesData';
import useUploadedFilesData from './useUploadedFilesData';

const stateKeys = StrictDict({
  hasSavedDraft: 'hasSavedDraft',
});

const useSubmissionViewData = () => {
  const [hasSavedDraft, setHasSavedDraft] = useKeyedState(stateKeys.hasSavedDraft, false);
  const hasSubmitted = useHasSubmitted();
  const setHasSubmitted = useSetHasSubmitted();
  const submitResponseMutation = useSubmitResponse();
  const rubricConfig = useRubricConfig();
  const globalState = useGlobalState({ step: stepNames.submission });
  const refreshPageData = useRefreshPageData();
  const setResponse = useSetResponse();
  const response = useResponse();
  const refreshUpstream = useRefreshUpstream();
  const stepState = hasSubmitted ? stepStates.submitted : globalState.stepState;

  const {
    textResponses,
    onUpdateTextResponse,
    saveResponse,
    saveResponseStatus,
    finishLater,
    finishLaterStatus,
  } = useTextResponsesData({ setHasSavedDraft });

  const {
    uploadedFiles,
    onFileUploaded,
    onDeletedFile,
  } = useUploadedFilesData();

  const submitResponseHandler = React.useCallback(() => {
    submitResponseMutation.mutateAsync({
      textResponses,
      uploadedFiles,
    }).then(() => {
      setResponse({ textResponses, uploadedFiles });
      setHasSubmitted(true);
      refreshPageData();
      refreshUpstream();
    });
  }, [ // eslint-disable-line react-hooks/exhaustive-deps
    setHasSubmitted,
    submitResponseMutation,
    textResponses,
    uploadedFiles,
  ]);

  return {
    actionOptions: {
      finishLater,
      finishLaterStatus,
      saveResponse,
      saveResponseStatus,
      submit: submitResponseHandler,
      submitStatus: submitResponseMutation.status,
      hasSubmitted,
    },
    response: hasSubmitted
      ? response
      : { textResponses, uploadedFiles },
    hasSubmitted,
    onUpdateTextResponse,
    isDraftSaved: hasSavedDraft,
    onDeletedFile,
    onFileUploaded,
    showRubric: rubricConfig.showDuringResponse,
    isReadOnly: stepState === stepStates.done || hasSubmitted,
  };
};

export default useSubmissionViewData;
