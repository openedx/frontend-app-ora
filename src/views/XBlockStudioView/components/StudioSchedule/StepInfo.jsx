import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';

import FormatDateTime from './FormatDateTime';

import messages from '../messages';

const StepInfo = ({ stepName, startDatetime, endDatetime }) => {
  const { formatMessage } = useIntl();
  return (
    <div>
      {startDatetime && (
        <p>
          <strong>{stepName} {formatMessage(messages.startLabel)}</strong>
          <FormatDateTime date={startDatetime} />
        </p>
      )}
      {endDatetime && (
        <p>
          <strong>{stepName} {formatMessage(messages.dueLabel)}</strong>
          <FormatDateTime date={endDatetime} />
        </p>
      )}
    </div>
  );
};

StepInfo.defaultProps = {
  startDatetime: null,
  endDatetime: null,
};

StepInfo.propTypes = {
  stepName: PropTypes.string.isRequired,
  startDatetime: PropTypes.string,
  endDatetime: PropTypes.string,
};

export default StepInfo;
