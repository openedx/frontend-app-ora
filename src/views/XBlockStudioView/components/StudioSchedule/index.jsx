import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Collapsible } from '@edx/paragon';
import { useORAConfigData } from 'hooks/app';

import StepInfo from './StepInfo';
import { useXBlockStudioViewContext } from '../XBlockStudioViewProvider';
import FormatDateTime from './FormatDateTime';

import messages from '../messages';

const StudioSchedule = () => {
  const { formatMessage } = useIntl();
  const { assessmentSteps, submissionConfig } = useORAConfigData();

  const { scheduleIsOpen, toggleSchedule } = useXBlockStudioViewContext();

  return (
    <Collapsible
      title={<h3 className="py-3">{formatMessage(messages.scheduleHeader)}</h3>}
      open={scheduleIsOpen}
      onToggle={toggleSchedule}
    >
      <div>
        <p>
          <strong>{formatMessage(messages.responseLabel)} {formatMessage(messages.startLabel)}</strong>
          <FormatDateTime date={submissionConfig.startDatetime} />
        </p>
        <p>
          <strong>{formatMessage(messages.responseLabel)} {formatMessage(messages.dueLabel)}</strong>
          <FormatDateTime date={submissionConfig.endDatetime} />
        </p>
        {Object.keys(assessmentSteps.settings).map((stepName) => (
          <StepInfo
            key={stepName}
            stepName={formatMessage(messages[stepName])}
            {...assessmentSteps.settings[stepName]}
          />
        ))}
      </div>
    </Collapsible>
  );
};

export default StudioSchedule;
