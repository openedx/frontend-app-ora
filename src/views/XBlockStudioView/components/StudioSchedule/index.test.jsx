import { shallow } from '@edx/react-unit-test-utils';

import { useORAConfigData } from 'hooks/app';
import { stepNames } from 'constants/index';

import StudioSchedule from './index';

jest.mock('hooks/app', () => ({
  useORAConfigData: jest.fn(),
}));
jest.mock('../XBlockStudioViewProvider', () => ({
  useXBlockStudioViewContext: () => ({
    scheduleIsOpen: true,
    toggleSchedule: jest.fn().mockName('toggleSchedule'),
  }),
}));
jest.mock('./FormatDateTime', () => 'FormatDateTime');
jest.mock('./StepInfo', () => 'StepInfo');

describe('<StudioSchedule />', () => {
  it('render without assesssment steps', () => {
    useORAConfigData.mockReturnValue({
      assessmentSteps: {
        settings: {},
      },
      submissionConfig: {
        startDatetime: '2020-01-01T00:00:00Z',
        endDatetime: '2020-01-01T00:00:00Z',
      },
    });

    const wrapper = shallow(<StudioSchedule />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('StepInfo')).toHaveLength(0);
  });

  it('render with assesssment steps', () => {
    useORAConfigData.mockReturnValue({
      assessmentSteps: {
        settings: {
          [stepNames.self]: {
            abc: 'def',
          },
          [stepNames.peer]: {
            ghi: 'jkl',
          },
        },
      },
      submissionConfig: {
        startDatetime: '2020-01-01T00:00:00Z',
        endDatetime: '2020-01-01T00:00:00Z',
      },
    });

    const wrapper = shallow(<StudioSchedule />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('StepInfo')).toHaveLength(2);
  });
});
