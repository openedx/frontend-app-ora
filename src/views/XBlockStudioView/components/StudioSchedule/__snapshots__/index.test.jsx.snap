// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudioSchedule /> render with assesssment steps 1`] = `
<Collapsible
  onToggle={[MockFunction toggleSchedule]}
  open={true}
  title={
    <h3
      className="py-3"
    >
      Schedule
    </h3>
  }
>
  <div>
    <p>
      <strong>
        Response
         
        start: 
      </strong>
      <FormatDateTime
        date="2020-01-01T00:00:00Z"
      />
    </p>
    <p>
      <strong>
        Response
         
        due: 
      </strong>
      <FormatDateTime
        date="2020-01-01T00:00:00Z"
      />
    </p>
    <StepInfo
      abc="def"
      key="self"
      stepName="Self assessment"
    />
    <StepInfo
      ghi="jkl"
      key="peer"
      stepName="Peer assessment"
    />
  </div>
</Collapsible>
`;

exports[`<StudioSchedule /> render without assesssment steps 1`] = `
<Collapsible
  onToggle={[MockFunction toggleSchedule]}
  open={true}
  title={
    <h3
      className="py-3"
    >
      Schedule
    </h3>
  }
>
  <div>
    <p>
      <strong>
        Response
         
        start: 
      </strong>
      <FormatDateTime
        date="2020-01-01T00:00:00Z"
      />
    </p>
    <p>
      <strong>
        Response
         
        due: 
      </strong>
      <FormatDateTime
        date="2020-01-01T00:00:00Z"
      />
    </p>
  </div>
</Collapsible>
`;
